Add support for OpenGL to OpenXR

dsnopek opened the pull request:
This PR aims to add support for the OpenGL renderer to OpenXR.

~~It's currently a DRAFT for a couple reasons:~~

- ~~It includes the changes from PR #65334 which needs to get merged first~~
- ~~There's only support for Windows - Linux and Android still need to be finished~~
- ~~Rather than updating the attachment on the render target's FBO each time it's overridden, I'd like to cache and reuse FBO's for each combination of color and depth texture (since OpenXR is swapping these every frame)~~
- ~~It currently renders upside down :-)~~

**UPDATE:** This is no longer a draft and is ready for review!

dsnopek committed:
Add support for OpenGL to OpenXR

BastiaanOlij commented on modules/openxr/extensions/openxr_opengl_extension.cpp:
```diff
@@ -0,0 +1,471 @@
+		OpenXRGraphicsExtensionWrapper(p_openxr_api) {
+#ifdef ANDROID
+	request_extensions[XR_KHR_OPENGL_ES_ENABLE_EXTENSION_NAME] = nullptr;
+	request_extensions[XR_KHR_ANDROID_THREAD_SETTINGS_EXTENSION_NAME] = nullptr;
```
`XR_KHR_ANDROID_THREAD_SETTINGS_EXTENSION_NAME` might need to go onto the android extension, surprised it's not there already..

dsnopek replied:
It is in the Android extension, so I've removed it from here

BastiaanOlij reviewed (approved):
This looks good to me, first merge the multiview PR then this one is good to go.

BastiaanOlij commented:
Re Linux build failures:
```
In file included from modules/openxr/extensions/openxr_vulkan_extension.h:57,
                 from modules/openxr/openxr_api.cpp:62:
thirdparty/openxr/include/openxr/openxr_platform.h:116:5: error: 'GLXFBConfig' does not name a type
  116 |     GLXFBConfig                 glxFBConfig;
      |     ^~~~~~~~~~~
thirdparty/openxr/include/openxr/openxr_platform.h:117:5: error: 'GLXDrawable' does not name a type; did you mean 'Drawable'?
  117 |     GLXDrawable                 glxDrawable;
      |     ^~~~~~~~~~~
      |     Drawable
thirdparty/openxr/include/openxr/openxr_platform.h:118:5: error: 'GLXContext' does not name a type; did you mean 'GContext'?
  118 |     GLXContext                  glxContext;
      |     ^~~~~~~~~~
      |     GContext
```

I remember running into this with the plugin too, vaguely recollect there was an additional include for opengl there?

BastiaanOlij commented:
The Android build looks like some of the defines may be wrong, it looks like it's taking the Linux code path, not the android one.

BastiaanOlij commented on modules/openxr/openxr_api.cpp:
```diff
@@ -45,10 +45,27 @@
+#define XR_USE_GRAPHICS_API_VULKAN
+#endif
+#ifdef GLES3_ENABLED
+#ifdef ANDROID
```
Indeed, the `ANDROID` define is now `ANDROID_ENABLED` so change this and in the other places.

dsnopek commented:
@BastiaanOlij Thanks for the review! I haven't actually started working on Linux or Android support yet, I had just copy-pasted the code from the OpenXR extension for Godot 3. :-) However, I think I did manage to get Linux compiling, but it still won't work quite yet. I'll come back to Linux and Android soon, and address all your comments then.

**General updates:** the changes I just pushed get it rendering on Windows (without the flashing that was happening previously), but it renders upside down. :-( Normally, we flip the image when blitting from the render target to the screen, but in this case we're rendering directly to the textures provided by OpenXR. I discussed some ideas with @clayjohn and @BastiaanOlij on RocketChat, but we're still not sure how to address it.

dsnopek commented:
Ok, I've added the remaining bits to theoretically make Android and Linux work, although, I haven't tested them (I don't really have a good way to test them, unfortunately). In doing so, I've addressed all of Bastiaan's review related to them!

The remaining tasks include:

- Fix the issue with it rendering upside down
- Rather than updating the attachment on the render target's FBO each time it's overridden, I'd like to cache and reuse FBO's for each combination of color and depth texture (since OpenXR is swapping these every frame)

bruvzg commented on servers/display_server.h:
```diff
@@ -70,6 +70,7 @@ class DisplayServer : public Object {
 		DISPLAY_HANDLE,
 		WINDOW_HANDLE,
 		WINDOW_VIEW,
+		OPENGL_CONTEXT,
```
New constant should be bound here:
https://github.com/godotengine/godot/blob/9ff3a43a329182ba2fad34263297052c70a4524e/servers/display_server.cpp#L834-L837

dsnopek replied:
Ah, thanks for catching that! Added in my latest push

bruvzg commented on platform/linuxbsd/display_server_x11.cpp:
```diff
@@ -1309,6 +1309,11 @@ int64_t DisplayServerX11::window_get_native_handle(HandleType p_handle_type, Win
 		case WINDOW_VIEW: {
 			return 0; // Not supported.
 		}
+#ifdef GLES3_ENABLED
+		case OPENGL_CONTEXT: {
+			return (int64_t)gl_manager->get_glx_context(p_window);
+		}
+#endif
```
For consistency, it's probably should be added to the macOS `DisplayServer` to return `NSOpenGLContext *` (even if it's not relevant for OpenXR).

dsnopek replied:
That makes sense! I don't have a good way to test it, but I've added some code that should do this

dsnopek commented:
My last push added caching of the FBO's for the sets of overridden textures, based on #68212, which'll hopefully be an easy one to review and merge.

This now just leaves figuring out the "it's rendering upside down" issue!

dsnopek commented:
I spent a few hours trying to see if it would be possible to just make the OpenGL renderer render right-side up all the time, since that would be good in non-XR as well, in order to render directly to the screen (rather than always rendering to a texture first, and then blitting that texture to the screen). However, after digging into it a bit, it was clear that that's going to be a pretty big endeavor!

So, instead, I've put a little hack into `rasterizer_scene_gles3.cpp` to not flip the Y axis if we are rendering to a render target that has it's color texture overridden. Hopefully, that'll be acceptable! Later, if we do manage to switch the renderer to render right-side up normally, we can remove these extra lines of code.

@BastiaanOlij @clayjohn But that means this PR is finally ready for a full review, because it is finally fully working :-)

BastiaanOlij commented on drivers/gles3/rasterizer_scene_gles3.cpp:
```diff
@@ -1729,8 +1732,20 @@ void RasterizerSceneGLES3::render_scene(const Ref<RenderSceneBuffers> &p_render_
+	}
+	if (!flip_y) {
+		// If we're rendering right-side up, then we need to change the winding order.
+		glFrontFace(GL_CW);
```
Should we set this back the other way around if flip_y is true? As far as I know this is a global setting so if we're rendering additional non-XR viewports to be used as textures, we'll end up rendering those wrong.

BastiaanOlij commented on drivers/gles3/rasterizer_scene_gles3.cpp:
```diff
@@ -1917,6 +1932,11 @@ void RasterizerSceneGLES3::render_scene(const Ref<RenderSceneBuffers> &p_render_
 
 	_render_list_template<PASS_MODE_COLOR_TRANSPARENT>(&render_list_params_alpha, &render_data, 0, render_list[RENDER_LIST_ALPHA].elements.size(), true);
 
+	if (!flip_y) {
```
Ah, you're unsetting it.. I guess thats another way to do it :)

BastiaanOlij commented on drivers/gles3/storage/render_scene_buffers_gles3.cpp:
```diff
@@ -33,17 +33,12 @@
 #include "render_scene_buffers_gles3.h"
 #include "texture_storage.h"
 
-#ifdef ANDROID_ENABLED
```
Did you mean to remove the multiview support? or is it moved somewhere else?

clayjohn replied:
Its not used now that depth is initialized in the render target allocation code

dsnopek replied:
Yeah, it's not removed completely - just moved from the render buffers over to the render targets!

BastiaanOlij replied:
Ah makes sense!

BastiaanOlij reviewed (approved):
This looks great to me, fine whichever way around we decide to go with `glFrontFace`, and I'm probably misreading the code removed for multiview, but just in case something was removed by accident.

clayjohn commented on modules/openxr/extensions/openxr_opengl_extension.cpp:
```diff
@@ -0,0 +1,480 @@
+	graphics_binding_gl.config = (EGLConfig)0; // https://github.com/KhronosGroup/OpenXR-SDK-Source/blob/master/src/tests/hello_xr/graphicsplugin_opengles.cpp#L122
+	graphics_binding_gl.context = eglGetCurrentContext();
+#else
+	graphics_binding_gl.type = XR_TYPE_GRAPHICS_BINDING_OPENGL_XLIB_KHR;
```
Is this X11 specific? I'm wondering what will happen if a MacOS device attempts to run this code

dsnopek replied:
Yes, this is code is X11-specific! AKAIK, OpenXR doesn't work on MacOS. I'll double check the build system and make sure this code isn't compiled on MacOS

clayjohn commented on modules/openxr/extensions/openxr_opengl_extension.cpp:
```diff
@@ -0,0 +1,480 @@
+	data->is_multiview = (p_array_size > 1);
+
+	// @todo How to get the real format?
+	Image::Format format = Image::FORMAT_RGBA8;
```
If you are referring to ``Texture.real_format`` in the GLES3 code, you use ``_get_gl_image_and_format()`` Which validates the texture type. ``real_format`` is used in case we need to decompress the texture (or convert it to a supported format). It represents what is actually stored on the GPU as opposed to the type of the original image. If you always use RGBA8, then you can just ignore it

clayjohn reviewed (commented):
Looks great so far!

clayjohn commented:
Great work!

dsnopek commented on modules/openxr/extensions/openxr_opengl_extension.cpp:
```diff
@@ -0,0 +1,479 @@
+	}
+	if (graphics_binding_gl.glxContext == nullptr) {
+		print_line("OpenXR Failed to get glxContext from Godot, using glXGetCurrentContext()");
+		graphics_binding_gl.glxContext = glXGetCurrentContext();
```
That's a good question! I copied this code from the OpenXR plugin for Godot 3. I think there may have been a time before `OS::get_native_handle()` (the name of the function in Godot 3) was added to Godot, and this code was a fallback for if you used the plugin with an older version of Godot. We should probably just remove all those fallbacks, since we know we can rely on `DisplayServer::window_get_native_handle()`. And now that in Godot 4 we have multiple window support, it's more likely that calling `glXGetCurrentContext()` would give us the incorrect context anyway

dsnopek replied:
Yeah, I found the "paper trail" for this!

PR #42531 added `OS::get_native_handle()` added to Godot 3, so that https://github.com/GodotVR/godot_openxr/pull/6 could use it in the OpenXR plugin.

I'll make a PR to remove the redundant fallback code...

dsnopek replied:
And here's the PR to remove it: https://github.com/godotengine/godot/pull/68453
